### PR TITLE
Rest api

### PR DESCRIPTION
**Pull request by quinarygio**
Rest API implementation

### CODE REVIEW

File: online-rest-api/src/test/java/eu/itesla_project/online/rest/api/test/ProcessApiTest.java [hunk 1]
```diff
@@ -0,0 +1,521 @@
...
+        public void storeWorkflowParameters(String workflowId, OnlineWorkflowParameters parameters) {
+
+        }
+
+        @Override
+        public OnlineWorkflowParameters getWorkflowParameters(String workflowId) {
+            DateTime dt = workflowMap.get(workflowId);
+            OnlineWorkflowParameters param = new OnlineWorkflowParameters(dt, 0, new Interval(0, 0),
+                    "OfflineworkflowId", TimeHorizon.DACF, workflowId, 0, false, false, false, null, CaseType.FO,
+                    new HashSet<Country>(), false, 0, false, 0);
```
Comment (mathbagu) [inline online-rest-api/src/test/java/eu/itesla_project/online/rest/api/test/ProcessApiTest.java]: Collections.emptySet() instead of new HashSet()

File: online-workflow/pom.xml [hunk 2]
```diff
@@ -87,6 +87,19 @@
...
+		    <groupId>com.fasterxml.jackson.core</groupId>
+		    <artifactId>jackson-annotations</artifactId>
+		    <version>2.4.1</version>
+		</dependency>
+		<dependency>
+			<groupId>joda-time</groupId>
+			<artifactId>joda-time</artifactId>
+		</dependency>
     </dependencies>
 </project>
```
Comment (mathbagu) [inline online-workflow/pom.xml]: Dependencies version should be defined in the parent pom

File: online-workflow/src/main/java/eu/itesla_project/online/LocalOnlineApplication.java [hunk 3]
```diff
@@ -64,65 +66,35 @@
...
-            throws IllegalAccessException, InstantiationException, InstanceAlreadyExistsException,
-            MBeanRegistrationException, MalformedObjectNameException, NotCompliantMBeanException, IOException {
+    public LocalOnlineApplication(OnlineConfig config, ComputationManager computationManager,
+            ScheduledExecutorService ses, ExecutorService es, boolean enableJmx)
+                    throws IllegalAccessException, InstantiationException, InstanceAlreadyExistsException,
+                    MBeanRegistrationException, MalformedObjectNameException, NotCompliantMBeanException, IOException {
         this.config = config;
         this.computationManager = computationManager;
+        this.ses = ses;
+        this.es = es;
```
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/LocalOnlineApplication.java]: Assert required parameters are not null

File: online-workflow/src/main/java/eu/itesla_project/online/LocalOnlineApplication.java [hunk 4]
```diff
@@ -211,52 +158,121 @@ public int getAvailableCores() {
         return computationManager.getResourcesStatus().getAvailableCores();
     }
 
+    @Override
+    public String startProcess(String name, String owner, DateTime date, DateTime creationDate,
+            OnlineWorkflowStartParameters start, OnlineWorkflowParameters params, DateTime[] basecases)
+                    throws Exception {
```
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/LocalOnlineApplication.java]: Assert that required parameters are not null

File: online-workflow/src/main/java/eu/itesla_project/online/LocalOnlineApplication.java [hunk 5]
```diff
@@ -211,52 +158,121 @@ public int getAvailableCores() {
...
+        proc.setOwner(owner);
+        proc.setCaseType(params.getCaseType().toString());
+        if (date != null)
+            proc.setDate(date);
+        else
+            proc.setDate(new DateTime());
+        if (creationDate != null)
+            proc.setCreationDate(creationDate);
+        else
+            proc.setCreationDate(new DateTime());
```
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/LocalOnlineApplication.java]: Call a constructor instead of setters

File: online-workflow/src/main/java/eu/itesla_project/online/LocalOnlineApplication.java [hunk 6]
```diff
@@ -383,54 +373,40 @@ public void onStatesWithIndexesUpdate(ContingencyStatesIndexesSynthesis indexeAc
...
+     * onUnstableContingencies(UnstableContingenciesSynthesis
+     * unstableContingencies) { sendNotification(new
+     * AttributeChangeNotification(this, notificationIndex.getAndIncrement(),
+     * System.currentTimeMillis(), "UnstableContingencies",
+     * UNSTABLE_CONTINGENCIES_ATTRIBUTE,
+     * unstableContingencies.getClass().getName(), null,
+     * unstableContingencies));
+     * 
+     * }
+     */
```
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/LocalOnlineApplication.java]: Could be removed?

File: online-workflow/src/main/java/eu/itesla_project/online/LocalOnlineApplication.java [hunk 7]
```diff
@@ -484,38 +458,36 @@ public void runFeaAnalysis(OnlineWorkflowStartParameters startconfig, ForecastEr
             e.printStackTrace();
             throw new RuntimeException(e);
         } finally {
-            workflowLock.unlock();
-            workflow = null;
+            // workflowLock.unlock();
+            // workflow=null;
```
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/LocalOnlineApplication.java]: Remove commented lines

File: online-workflow/src/main/java/eu/itesla_project/online/OnlineApplication.java [hunk 8]
```diff
@@ -25,5 +27,9 @@
     void removeListener(OnlineApplicationListener l);
     
     void notifyListeners();
+    
+    String startProcess(String name, String owner, DateTime date, DateTime creationDate,
+			OnlineWorkflowStartParameters start, OnlineWorkflowParameters params, DateTime[] basecases)
+					throws Exception;
```
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/OnlineApplication.java]: Fix indentation

File: online-workflow/src/main/java/eu/itesla_project/online/OnlineWorkflow.java [hunk 9]
```diff
@@ -11,7 +11,11 @@
  * @author Quinary <itesla@quinary.com>
  */
 public interface OnlineWorkflow {
-
+	/*
+	String getProcessId();
+	
+	void setProcessId(String processId);
+	*/
```
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/OnlineWorkflow.java]: Remove commented lines
Fix indentation

File: online-workflow/src/main/java/eu/itesla_project/online/OnlineWorkflowImpl.java [hunk 10]
```diff
@@ -70,7 +70,7 @@
     private final MergeOptimizerFactory mergeOptimizerFactory;
     private final RulesFacadeFactory rulesFacadeFactory;
     private final OnlineWorkflowStartParameters startParameters;
-
+    //private String processId;
```
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/OnlineWorkflowImpl.java]: Remove commented lines

File: online-workflow/src/main/java/eu/itesla_project/online/OnlineWorkflow.java [hunk 11]
```diff
@@ -20,4 +24,6 @@
 
 	void removeOnlineApplicationListener(OnlineApplicationListener listener);
 
+	
+
```
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/OnlineWorkflow.java]: Remove extra empty lines

File: online-workflow/src/main/java/eu/itesla_project/online/OnlineWorkflowImpl.java [hunk 12]
```diff
@@ -137,8 +137,20 @@ public OnlineWorkflowImpl(
...
+    	this.processId=processId;
+    }
+    
+    @Override
+    public String getProcessId(){
+    	return processId;
+    }
+    */
+    
+
```
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/OnlineWorkflowImpl.java]: Remove commented lines

File: online-workflow/src/main/java/eu/itesla_project/online/RemoteOnlineApplication.java [hunk 13]
```diff
@@ -182,13 +184,14 @@ public int getAvailableCores() {
...
+
     @Override
     public String startWorkflow(OnlineWorkflowStartParameters start, OnlineWorkflowParameters params) {
         try {
             return application.startWorkflow(start, params);
-        } catch (Exception e) {
+        } catch (Throwable e) {
             LOGGER.error(e.getMessage(),e);
-            notifyDisconnection();
+           // notifyDisconnection();
```
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/RemoteOnlineApplication.java]: Remove commented lines

File: online-workflow/src/main/java/eu/itesla_project/online/RemoteOnlineApplication.java [hunk 14]
```diff
@@ -227,4 +230,19 @@ public void removeListener(OnlineApplicationListener l) {
...
+					throws Exception {
+		try {
+        	return application.startProcess( name,  owner,  date,  creationDate, start,  params,  basecases);            
+        } catch (Exception  e) {
+        	LOGGER.error(e.getMessage(),e);
+        	notifyDisconnection();
+        }
+		return null;
+	}
+
```
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/RemoteOnlineApplication.java]: Fix indentation + extra spaces

File: online-workflow/src/main/java/eu/itesla_project/online/StatusSynthesis.java [hunk 15]
```diff
@@ -24,6 +24,7 @@
 	public static int STATUS_RUNNING=1;
 	public static int STATUS_STOPPED=2;
 	public static int STATUS_TERMINATED=3;
+	public static int STATUS_ERROR=4;
```
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/StatusSynthesis.java]: Could be converted to an enum?

File: online-workflow/src/main/java/eu/itesla_project/online/tools/OnlineProcessParameters.java [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+*/
+public class OnlineProcessParameters{
+
+	private String caseType;
+	private Integer states;
+	private String name;
+	private String owner;
+	private DateTime creationDate;
+	private DateTime date;
+	private String basecasesInterval;
```
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/tools/OnlineProcessParameters.java]: Verify if these parameters could become final
Verify if these parameters could really be null

File: online-workflow/src/main/java/eu/itesla_project/online/tools/OnlineProcessTool.java [hunk 17]
```diff
@@ -0,0 +1,303 @@
...
+       	else
+       		System.out.println("File not found: "+filename);
+		return null;
+	}
+
+
+	
+	
+
+}
```
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/tools/OnlineProcessTool.java]: Fix coding style

File: pom.xml [hunk 18]
```diff
@@ -50,6 +50,7 @@
         <module>online-mpi</module>
         <module>online-server</module>
         <module>online-workflow</module>
+		<module>online-rest-api</module>
```
Comment (mathbagu) [inline pom.xml]: Fix indentation

File: online-workflow/src/main/java/eu/itesla_project/online/tools/OnlineProcessParameters.java (hunk 16 above)
Comment (quinarygio) [inline online-workflow/src/main/java/eu/itesla_project/online/tools/OnlineProcessParameters.java]: Cannot be final because in the OnlineProcessTool they are first read from config-file but can be overwritten by command-line options.

File: online-mpi/src/main/java/eu/itesla_project/online/mpi/Master.java [hunk 19]
```diff
@@ -82,13 +82,13 @@ public static void main(String[] args) throws Exception {
 
             MpiExecutorContext mpiExecutorContext = new MultiStateNetworkAwareMpiExecutorContext();
             ScheduledExecutorService scheduledExecutorService = Executors.newScheduledThreadPool(1);
-            ExecutorService executorService = MultiStateNetworkAwareExecutors.newCachedThreadPool();
+            //ExecutorService executorService = MultiStateNetworkAwareExecutors.newCachedThreadPool();
```
Comment (geofjamg) [inline online-mpi/src/main/java/eu/itesla_project/online/mpi/Master.java]: Why network aware executor is not used anymore? It multi-state IIDM network used inside the online process? network aware executor ensure current state in correctly set when getting data from a network in a new thread.

File: online-rest-api/src/main/java/eu/itesla_project/online/rest/api/util/OnlineDBUtils.java [hunk 20]
```diff
@@ -0,0 +1,254 @@
...
+                    && onlinedb.getWorkflowDetails(workflowId) != null) {
+
+                OnlineWorkflowParameters params = onlinedb.getWorkflowParameters(workflowId);
+                WorkflowResult res = new WorkflowResult(processId, workflowId,
+                        params.getBaseCaseDate().toDateTimeISO().toString());
+
+                Map<Integer, List<LimitViolation>> violations = onlinedb.getViolations(workflowId,
+                        OnlineStep.LOAD_FLOW);
+                if (violations != null) {
+                    violations.forEach(new BiConsumer<Integer, List<LimitViolation>>() {
```
Comment (geofjamg) [inline online-rest-api/src/main/java/eu/itesla_project/online/rest/api/util/OnlineDBUtils.java]: Lambda?

File: online-rest-api/src/main/java/eu/itesla_project/online/rest/api/util/OnlineDBUtils.java [hunk 21]
```diff
@@ -0,0 +1,254 @@
...
+                            res.addPreContingency(pcr);
+                        }
+                    });
+                }
+
+                Map<String, Map<Integer, SimulationResult>> postMap = new HashMap<String, Map<Integer, SimulationResult>>();
+
+                Map<Integer, Map<String, Boolean>> conv = onlinedb.getPostContingencyLoadflowConvergence(workflowId);
+                if (conv != null) {
+                    conv.forEach(new BiConsumer<Integer, Map<String, Boolean>>() {
```
Comment (geofjamg) [inline online-rest-api/src/main/java/eu/itesla_project/online/rest/api/util/OnlineDBUtils.java]: lambda?

File: online-rest-api/src/main/java/eu/itesla_project/online/rest/api/util/OnlineDBUtils.java [hunk 22]
```diff
@@ -0,0 +1,254 @@
...
+                                sr.setConvergence(convergenceMap.get(cont));
+                            }
+                        }
+                    });
+                }
+
+                Map<Integer, Map<String, List<LimitViolation>>> violsMap = onlinedb
+                        .getPostContingencyViolations(workflowId);
+                if (violsMap != null) {
+                    violsMap.forEach(new BiConsumer<Integer, Map<String, List<LimitViolation>>>() {
```
Comment (geofjamg) [inline online-rest-api/src/main/java/eu/itesla_project/online/rest/api/util/OnlineDBUtils.java]: Lambda?

File: online-rest-api/src/main/java/eu/itesla_project/online/rest/api/util/OnlineDBUtils.java [hunk 23]
```diff
@@ -0,0 +1,254 @@
...
+                            if (sr == null) {
+                                sr = new SimulationResult(stateId);
+                                srMap.put(stateId, sr);
+                            }
+                            sr.setSafe(!wfResults.getIndexesData(contingencyId, stateId).containsValue(Boolean.FALSE));
+                        }
+                    }
+                }
+
+                postMap.forEach(new BiConsumer<String, Map<Integer, SimulationResult>>() {
```
Comment (geofjamg) [inline online-rest-api/src/main/java/eu/itesla_project/online/rest/api/util/OnlineDBUtils.java]: Lambda?

File: online-rest-api/src/main/java/eu/itesla_project/online/rest/api/util/OnlineDBUtils.java [hunk 24]
```diff
@@ -0,0 +1,254 @@
...
+            LOGGER.error(e.getMessage(), e);
+            throw new Exception(e.getMessage());
+        }
+        return null;
+    }
+
+    private Process toProcess(OnlineProcess p) {
+        Process proc = new Process(p.getId(), p.getName(), p.getOwner(), p.getDate().toDate(),
+                p.getCreationDate().toDate());
+        p.getWorkflowsMap().forEach(new BiConsumer<String, String>() {
```
Comment (geofjamg) [inline online-rest-api/src/main/java/eu/itesla_project/online/rest/api/util/OnlineDBUtils.java]: Lambda?

File: online-rest-api/src/main/java/eu/itesla_project/online/rest/model/PreContingencyResult.java [hunk 25]
```diff
@@ -0,0 +1,103 @@
...
+public class PreContingencyResult {
+
+    private final Integer state;
+    private final Boolean convergence;
+    private final Boolean safe;
+    private final List<Violation> violations = new ArrayList<>();
+
+    public PreContingencyResult(Integer state, boolean safe, boolean convergence) {
+        this.state = Objects.requireNonNull(state);
+        this.safe = Objects.requireNonNull(safe);
```
Comment (geofjamg) [inline online-rest-api/src/main/java/eu/itesla_project/online/rest/model/PreContingencyResult.java]: requireNonNull cannot be used on a primitive type

File: online-rest-api/src/main/java/eu/itesla_project/online/rest/model/PreContingencyResult.java [hunk 26]
```diff
@@ -0,0 +1,103 @@
...
+        this.convergence = Objects.requireNonNull(convergence);
```
Comment (geofjamg) [inline online-rest-api/src/main/java/eu/itesla_project/online/rest/model/PreContingencyResult.java]: same

File: online-rest-api/src/main/java/eu/itesla_project/online/rest/model/PreContingencyResult.java [hunk 27]
```diff
@@ -0,0 +1,103 @@
...
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(state, convergence, safe, violations);
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("class PreContingencyResult {\n");
```
Comment (geofjamg) [inline online-rest-api/src/main/java/eu/itesla_project/online/rest/model/PreContingencyResult.java]: General comment: isn't \n too platform dependent? System.lineSeparator() ?

File: online-server/pom.xml [hunk 28]
```diff
@@ -38,18 +35,16 @@
             <groupId>com.google.code.gson</groupId>
```
Comment (geofjamg) [inline online-server/pom.xml]: Why gson and not jackson to be consistent everywhere?

File: online-rest-api/src/test/java/eu/itesla_project/online/rest/api/test/ProcessApiTest.java [hunk 29]
```diff
@@ -0,0 +1,718 @@
...
+        }
+
+        @Override
+        public Network getState(String workflowId, Integer stateId, String contingencyId) {
+            return null;
+        }
+
+    }
+
+}
```
Comment (mathbagu) [inline online-rest-api/src/test/java/eu/itesla_project/online/rest/api/test/ProcessApiTest.java]: The mock classes could be extracted to independant class files to limit the file size and make it easier to read

File: online-rest-api/src/test/resources/log4j.properties [hunk 30]
```diff
@@ -0,0 +1,9 @@
+# Set root logger level to DEBUG and its only appender to A1.
+log4j.rootLogger=INFO, A1
+
+# A1 is set to be a ConsoleAppender.
+log4j.appender.A1=org.apache.log4j.ConsoleAppender
+
+# A1 uses PatternLayout.
+log4j.appender.A1.layout=org.apache.log4j.PatternLayout
+log4j.appender.A1.layout.ConversionPattern=%-4r [%t] %-5p %c %x - %m%n
```
Comment (mathbagu) [inline online-rest-api/src/test/resources/log4j.properties]: Why cannot use log4j-over-slf4j to avoid having 2 logging system is the same software.

File: online-server/pom.xml [hunk 31]
```diff
@@ -38,18 +35,16 @@
...
             <artifactId>gson</artifactId>
         </dependency>
-
         <dependency>
             <groupId>javax</groupId>
             <artifactId>javaee-api</artifactId>
             <version>7.0</version>
```
Comment (mathbagu) [inline online-server/pom.xml]: Use a variable from main pom.xml

File: online-server/src/main/java/eu/itesla_project/online/server/OnlineApplicationBean.java [hunk 32]
```diff
@@ -57,285 +57,220 @@
...
-	{
-		return workflows;		
-		
-	}
-	
-	public boolean isJmxConnected(){
-		return (remoteConnected && application !=null );
-	}
-	
-
```
Comment (mathbagu) [inline online-server/src/main/java/eu/itesla_project/online/server/OnlineApplicationBean.java]: A lot of input parameters could be null and cause NullPointerException. It could be interested to add assertion to test where it's needed

File: online-server/src/main/java/eu/itesla_project/online/server/OnlineApplicationResource.java [hunk 33]
```diff
@@ -82,310 +79,247 @@ public void notifyListeners() {
...
+        }
+
+        HttpSession session = req.getSession();
+        if (session != null)
+            try {
+
+                session.invalidate();
+
+            } catch (Exception ex) {
+            }
```
Comment (mathbagu) [inline online-server/src/main/java/eu/itesla_project/online/server/OnlineApplicationResource.java]: Coding style is not valid

File: online-server/src/main/java/eu/itesla_project/online/server/OnlineWorkflowInfo.java [hunk 34]
```diff
@@ -15,117 +14,87 @@
...
+    private WorkflowStatusEnum status = WorkflowStatusEnum.UNDEFINED;
+    private boolean wcaRunning;
+    private HashMap<Integer, WorkStatus> workStatus;
+    private ContingencyStatesActionsSynthesis statesActions;
+    private ContingencyStatesIndexesSynthesis unsafeContingencies;
+    private IndexSecurityRulesResultsSynthesis securityRulesIndexesApplication;
+
+    private WcaContingenciesSynthesis wcaContingencies;
+
+    public OnlineWorkflowInfo(String id) {
```
Comment (mathbagu) [inline online-server/src/main/java/eu/itesla_project/online/server/OnlineWorkflowInfo.java]: Assert id is not null

File: online-server/src/main/java/eu/itesla_project/online/server/OnlineWorkflowInfo.java [hunk 35]
```diff
@@ -15,117 +14,87 @@
...
+
+    public String getWorflowId() {
+        return worflowId;
+    }
+
+    public WorkflowStatusEnum getStatus() {
+        return status;
+    }
+
+    public void setStatus(WorkflowStatusEnum workflowStatusEnum) {
```
Comment (mathbagu) [inline online-server/src/main/java/eu/itesla_project/online/server/OnlineWorkflowInfo.java]: Assert workflowStatusEnum is not null

File: online-server/src/main/java/eu/itesla_project/online/server/OnlineWorkflowInfo.java [hunk 36]
```diff
@@ -15,117 +14,87 @@
...
+        this.securityRulesIndexesApplication = indexesSecurityRulesApplicatio;
+    }
+
+    public WcaContingenciesSynthesis getWcaContingencies() {
+        return wcaContingencies;
+    }
+
+    public void setWcaContingencies(WcaContingenciesSynthesis wcaContingencies) {
+        this.wcaContingencies = wcaContingencies;
+    }
```
Comment (mathbagu) [inline online-server/src/main/java/eu/itesla_project/online/server/OnlineWorkflowInfo.java]: Assert input parameters are not null

File: online-server/src/main/java/eu/itesla_project/online/server/message/ConnectionMessage.java [hunk 37]
```diff
@@ -15,29 +15,27 @@
...
-
-		@Override
-		protected void toJson(JsonGenerator generator) {
-			
-
-		}
+
+    private String type = "connection";
+
+    public ConnectionMessage(Boolean body) {
```
Comment (mathbagu) [inline online-server/src/main/java/eu/itesla_project/online/server/message/ConnectionMessage.java]: Could be a boolean with auto-boxing?

File: online-server/src/main/java/eu/itesla_project/online/server/message/RunningMessage.java [hunk 38]
```diff
@@ -15,29 +15,27 @@
...
-		@Override
-		protected void toJson(JsonGenerator generator) {
-			
-
-		}
+
+    private String type = "running";
+
+    public RunningMessage(Boolean body) {
+        super(body);
```
Comment (mathbagu) [inline online-server/src/main/java/eu/itesla_project/online/server/message/RunningMessage.java]: Convert to boolean with auto-boxing?

File: online-server/src/main/webapp/index.html [hunk 39]
```diff
@@ -9,91 +9,98 @@
...
+		</div>
+	</div>
+	<div ng-if="currentUser != null">
+		<div class="container-fluid">
+			<div ng-view></div>
+		</div>
+	</div>
+
+</body>
 </html>
```
Comment (mathbagu) [inline online-server/src/main/webapp/index.html]: Replace tabs by spaces

File: online-server/src/main/webapp/index.xhtml [hunk 40]
```diff
@@ -10,18 +10,22 @@
...
-<h3><h:outputText value="#{msg.welcomeHeading}" /></h3>
-<p><h:outputText value="#{msg.welcomeMessage}" /></p>
+	<h3>
+		<h:outputText value="#{msg.welcomeHeading}" />
+	</h3>
+	<p>
+		<h:outputText value="#{msg.welcomeMessage}" />
+	</p>
 </body>
 </html>
```
Comment (mathbagu) [inline online-server/src/main/webapp/index.xhtml]: Replace tabs by spaces

File: online-server/src/main/webapp/loggedOut.html [hunk 41]
```diff
@@ -9,38 +9,42 @@
...
+	<div class="container login">
+		<h2 class="form-signin-heading-color">Logged Out</h2>
+
+		<div>
+			<a href="index.html">Login</a>
+		</div>
+	</div>
+
 </body>
 </html>
```
Comment (mathbagu) [inline online-server/src/main/webapp/loggedOut.html]: Replace tabs by spaces

File: online-server/src/main/webapp/login.html [hunk 42]
```diff
@@ -9,44 +9,50 @@
...
+				placeholder="Username" required autofocus> <input
+				type="password" id="j_password" name="j_password"
+				class="form-control" placeholder="Password" required>
+			<button class="btn btn-lg btn-primary btn-block btn-color"
+				type="submit">Login</button>
+
+		</form>
+	</div>
 </body>
 </html>
```
Comment (mathbagu) [inline online-server/src/main/webapp/login.html]: Replace tabs by spaces

File: online-server/src/main/webapp/loginError.html [hunk 43]
```diff
@@ -9,33 +9,36 @@
...
+	<div class="container login">
+		<h2 class="form-signin-heading-color">Login failed</h2>
+		Invalid username or password.
+		<div>
+			<a href="index.html">Retry login</a>
+		</div>
+	</div>
+
 </body>
 </html>
```
Comment (mathbagu) [inline online-server/src/main/webapp/loginError.html]: Replace tabs by spaces

File: online-server/src/main/webapp/partials/computationplatform.html [hunk 44]
```diff
@@ -7,19 +7,19 @@
...
+	</div>
+	<div class="panel panel-default">
+		<div class="panel-heading">
+			<h3 class="panel-title">Real time monitoring</h3>
+		</div>
+		<div class="panel-body">
+			<real-time-area-chart width="960" height="500" points="{{busyCores}}" />
+		</div>
+	</div>
 </div>
```
Comment (mathbagu) [inline online-server/src/main/webapp/partials/computationplatform.html]: Replace tabs by spaces

File: online-server/src/main/webapp/partials/forecastResults.html [hunk 45]
```diff
@@ -7,21 +7,20 @@
...
+	<div class="panel panel-default">
+		<div class="panel-heading">
+			<h3 class="panel-title">titolo forecast result</h3>
+		</div>
+		<div class="panel-body">
+			<li ng-repeat="forecastResult in forecastResults">
+				{{forecastResult.name}} - {{forecastResult.value}}</li>
+		</div>
+	</div>
 </div>
```
Comment (mathbagu) [inline online-server/src/main/webapp/partials/forecastResults.html]: Replace tabs by spaces

File: online-server/src/main/webapp/partials/onlineWorkflow.html [hunk 46]
```diff
@@ -8,328 +8,468 @@
...
+									</tr>
+									</tr>
+								</tbody>
+							</table>
+						</td>
+					</tr>
+				</tbody>
+			</table>
+		</div>
+	</div>
```
Comment (mathbagu) [inline online-server/src/main/webapp/partials/onlineWorkflow.html]: Replace tabs by spaces

File: online-server/src/main/webapp/partials/workflows.html [hunk 47]
```diff
@@ -7,53 +7,63 @@
...
+					<td ng-class="{  running: workflow.status == 'RUNNING'  }"><h4
+							ng-class="{  running: workflow.status == 'RUNNING'  }">{{workflow.status
+							}}</h4></td>
+				</tr>
+			</tbody>
+		</table>
+
+	</div>
+
+</div>
```
Comment (mathbagu) [inline online-server/src/main/webapp/partials/workflows.html]: Replace tabs by spaces

File: online-workflow/src/main/java/eu/itesla_project/online/OnlineWorkflow.java [hunk 48]
```diff
@@ -20,4 +24,6 @@
 
 	void removeOnlineApplicationListener(OnlineApplicationListener listener);
```
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/OnlineWorkflow.java]: This method is never called: why don't you remove previously added listeners.

File: online-workflow/src/main/java/eu/itesla_project/online/StatusSynthesis.java (hunk 15 above)
Comment (mathbagu) [inline online-workflow/src/main/java/eu/itesla_project/online/StatusSynthesis.java]: Fix indentation

File: online-server/pom.xml (hunk 28 above)
Comment (quinarygio) [inline online-server/pom.xml]: There is already a general open issue on this matter, it will be managed in a separate pull request

File: online-rest-api/src/main/java/eu/itesla_project/online/rest/api/ApiOriginFilter.java [hunk 49]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ * @author Giovanni Ferrari <giovanni.ferrari@techrain.it>
+ */
+public class ApiOriginFilter implements javax.servlet.Filter {
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+            throws IOException, ServletException {
+
+        HttpServletResponse res = (HttpServletResponse) response;
```
Comment (quinarygio) [inline online-rest-api/src/main/java/eu/itesla_project/online/rest/api/ApiOriginFilter.java]: The doFilter method of the Filter is only called by the container when processing a request/response.
Do you want me to check for nulls anyway ?

File: online-rest-api/src/main/java/eu/itesla_project/online/rest/api/impl/ProcessApiServiceImpl.java [hunk 50]
```diff
@@ -0,0 +1,94 @@
...
+        this.utils = Objects.requireNonNull(utils);
+        objectMapper = new ObjectMapper();
+        objectMapper.registerModule(new JodaModule());
+        objectMapper.configure(com.fasterxml.jackson.databind.SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false);
+        objectMapper.setDateFormat(sdf);
+    }
+
+    @Override
+    public Response getProcessList(String owner, String basecase, String name, DateTimeParameter date,
+            DateTimeParameter creationDate, SecurityContext securityContext) {
```
Comment (quinarygio) [inline online-rest-api/src/main/java/eu/itesla_project/online/rest/api/impl/ProcessApiServiceImpl.java]: The SecurityContext is injected by the container. 
The other parameters are nullable by design, it should be possible to filer the process list only on a subset of the parameters.

File: online-rest-api/src/main/java/eu/itesla_project/online/rest/api/ProcessApiService.java [hunk 51]
```diff
@@ -0,0 +1,23 @@
...
+package eu.itesla_project.online.rest.api;
+
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.SecurityContext;
+
+/**
+ *
+ * @author Giovanni Ferrari <giovanni.ferrari@techrain.it>
+ */
+public abstract class ProcessApiService {
```
Comment (quinarygio) [inline online-rest-api/src/main/java/eu/itesla_project/online/rest/api/ProcessApiService.java]: It was automatically generated starting from a "swagger" definition of the REST API .

File: online-rest-api/src/main/java/eu/itesla_project/online/rest/api/impl/ProcessApiServiceImpl.java (hunk 50 above)
Comment (mathbagu) [inline online-rest-api/src/main/java/eu/itesla_project/online/rest/api/impl/ProcessApiServiceImpl.java]: Fine

File: online-server/pom.xml (hunk 28 above)
Comment (mathbagu) [inline online-server/pom.xml]: It's Ok for the existing modules which will be managed by the open issue. Ensure that the new online Rest module does not depend on gson
